src/watershed.jl

```diff
@@ -0,0 +1,97 @@
+import Base.isless
+
+immutable PixelKey{CT}
+    val::CT
+    time_step::Int
+end
+isless{T}(a::PixelKey{T}, b::PixelKey{T}) = (a.val < b.val) || (a.val == b.val && a.time_step < b.time_step)
+
+"""
+```
+segments                = watershed(img, [markers])
+``` 
+Segments the image using watershed transform. Each basin formed by watershed transform corresponds to a segment.
+If no markers are provided, local minima of the image are taken as markers.
+    
+Parameters:  
+-    img            = input grayscale image
+-    markers        = an array marking the pixels where flooding should start
+
+"""
+function watershed{T<:Images.NumberLike}(img::Array{T, 2}, markers::Array{T,2})
+
+    segments = copy(markers)
+    pq = PriorityQueue(CartesianIndex{2}, PixelKey{T}, Base.Order.Forward)
+    time_step = 0
+
+    R = CartesianRange(size(img))
+    Istart, Iend = first(R), last(R)
+    for i in CartesianRange(size(img))
+        if markers[i] > 0
+            for j in CartesianRange(max(Istart,i-1), min(i+1,Iend))
+                if segments[j] == 0
+                    segments[j] = markers[i]
+                    enqueue!(pq, j, PixelKey(img[i], time_step))
+                    time_step += 1
+                end
+            end
+        end
+    end
+
+    while length(pq) > 0
+        current = dequeue!(pq)
+        for j in CartesianRange(max(Istart,current-1), min(current+1,Iend))
+            if segments[j] == 0
+                segments[j] = segments[current]
+                enqueue!(pq, j, PixelKey(img[current], time_step))
+                time_step += 1
+            end
+        end
+    end
+
+    num_segments        = Int(maximum(segments))
+    labels              = Array(1:num_segments)
+    region_means        = Dict{Int, Images.accum(T)}()
+    region_pix_count    = Dict{Int, Int}()
+
+    for j in indices(img, 2)
+        for i in indices(img, 1)
+            region_pix_count[segments[i,j]] = get(region_pix_count, segments[i, j], 0) + 1
+            region_means[segments[i,j]] = get(region_means, segments[i,j], zero(Images.accum(T))) + (img[i, j] - get(region_means, segments[i,j], zero(Images.accum(T))))/region_pix_count[segments[i,j]]
+        end
+    end
+    return SegmentedImage(segments, labels, region_means, region_pix_count)
+end
+
+function watershed{T<:Images.NumberLike}(img::Array{T, 2})
+    markers = zeros(img)
+    i = 0
+    for i in findlocalmaxima(accumulator_matrix)
+        i += 1
+        markers[i] = i
```

Comment:
If `markers` is constrained to have the same eltype as `img`, won't you run into problems here for `::Matrix{N0f8}`?